### PR TITLE
emscripten support

### PR DESCRIPTION
**Pull request by lemire**
This is closely related to https://github.com/ada-url/ada/pull/411 but I am trying to keep everything as automated as possible.

You can build and test as follows:

```
emcmake  cmake -B buildwasm
cmake --build buildwasm
ctest --test-dir buildwasm
```

It is essentially a derivative of the work @MoLow did and he should get most of the credit. I just automated the process.

### CODE REVIEW

File: .github/workflows/emscripten.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    steps:
+      - uses: mymindstorm/setup-emsdk@v12
+      - name: Verify
+        run: emcc -v
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Configure
+        run: emcmake  cmake -B buildwasm
+      - name: Build
+        run: cmake --build buildwasm
```
Comment (anonrig) [inline .github/workflows/emscripten.yml:36]: I think we should test it the same way @molow does it

File: .github/workflows/emscripten.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: mymindstorm/setup-emsdk@v12
+      - name: Verify
+        run: emcc -v
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (anonrig) [inline .github/workflows/emscripten.yml:32]: This needs to be v3

File: .github/workflows/emscripten.yml (hunk 1 above)
Comment (lemire) [reply .github/workflows/emscripten.yml]: Done.

File: .github/workflows/emscripten.yml (hunk 2 above)
Comment (lemire) [reply .github/workflows/emscripten.yml]: Done.

**Comment by lemire** [comment]
@MoLow  What do you think? It appears that we pass your test without touching much of the code?

All I did was to add a few casts because emscripten is, for the time being, 32-bit...

**Comment by MoLow** [comment]
> It appears that we pass your test without touching much of the code?

this test uses the cli which does not use wasm bindings at all, just reads from stdout

**Comment by lemire** [comment]
> this test uses the cli which does not use wasm bindings at all, just reads from stdout

Ah.

**Comment by lemire** [comment]
@MoLow Verified.

**Comment by lemire** [comment]
@anonrig @MoLow We had a fun bug. It wasn't a problem with emscripten after all.

I have tweaked the last commit so that it credits @MoLow

File: CMakeLists.txt [hunk 3]
```diff
@@ -59,17 +63,19 @@ set_target_properties(
   SOVERSION "${ADA_LIB_SOVERSION}"
   WINDOWS_EXPORT_ALL_SYMBOLS YES
 )
-
-include (TestBigEndian)
-TEST_BIG_ENDIAN(IS_BIG_ENDIAN)
-if(IS_BIG_ENDIAN)
-  message(STATUS "Big-endian system detected.")
+if(NOT EMSCRIPTEN)
```
Comment (MoLow) [inline CMakeLists.txt:66]: this is not needed if `CMAKE_MODULE_PATH` is not set

**Comment by MoLow** [comment]
https://github.com/ada-url/ada/actions/runs/5007793980/jobs/8974870565?pr=412
```
Test project /home/runner/work/ada/ada
No tests were found!!!
```

**Comment by lemire** [comment]
> No tests were found!!! 

It probably fails to find node...?

**Comment by lemire** [comment]
@MoLow Ah no.


I had...

```
ctest --tests-dir buildwasm
```

It needs to be...
```
ctest --test-dir buildwasm
```

File: CMakeLists.txt (hunk 3 above)
Comment (lemire) [reply CMakeLists.txt]: Point taken but we don't need this big endian stuff anyhow.

File: .github/workflows/emscripten.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      - name: Verify
+        run: emcc -v
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Configure
+        run: emcmake  cmake -B buildwasm
+      - name: Build
+        run: cmake --build buildwasm
+      - name: Test
+        run: ctest --test-dir buildwasm
```
Comment (MoLow) [inline .github/workflows/emscripten.yml:39]: why don't we run the js test file directly? it is hard to analyze the output this way

Comment (MoLow) [reply .github/workflows/emscripten.yml:39]: besides that, LGTM

**Review by anonrig — APPROVED** [review]

File: .github/workflows/emscripten.yml (hunk 4 above)
Comment (anonrig) [reply .github/workflows/emscripten.yml:39]: I kind of feel we should do it directly

**Review by MoLow — APPROVED** [review]

**Pull request merged**